ddtrace/tracer/spancontext.go

```diff
@@ -144,13 +145,16 @@ const (
 // priority, the root reference and a buffer of the spans which are part of the
 // trace, if these exist.
 type trace struct {
-	mu               sync.RWMutex     // guards below fields
-	spans            []*span          // all the spans that are part of this trace
-	finished         int              // the number of finished spans
-	full             bool             // signifies that the span buffer is full
-	priority         *float64         // sampling priority
-	locked           bool             // specifies if the sampling priority can be altered
-	samplingDecision samplingDecision // samplingDecision indicates whether to send the trace to the agent.
+	mu               sync.RWMutex      // guards below fields
+	spans            []*span           // all the spans that are part of this trace
+	tags             map[string]string // trace level tags
+	propagatedTags   map[string]string // tags propagated between services
+	upstreamServices string            // _dd.p.upstream_services value from the upstream service
```

Comment:
Let's remove this too and remember the following:
* service name (probably already have it on the first span, no need to remember it)
* sampling rate
* sampler name

At the very end, we can modify `tags["_dd.p...."]` and append whatever the final decision is, upon flushing. That way we avoid parsing that string more than once.

Reply:
The solution you suggested improves the performance case by case.

1. If `Service A` changes its priority multiple times, and doesn't call any downstream service, then your suggestion works.
2. But, if `Service A` changes its priority once or twice, and calls `Service B` many times, then we need to re-construct `_dd.p.upstream_services` every time it calls `Service B`.

I cannot speak on behalf of others. But, in our case, I think `2` case happens more often.